Moved optimization script wizard to new dialog

robbr48 opened the pull request:
Resolves #1734 

Script generation is now completely separated from the optimization dialog. This should give a much more intuitive user experience, since you don't need to step through the generator pages before starting the script.

peterNordin commented on HopsanGUI/Dialogs/OptimizationScriptWizard.cpp:
```diff
@@ -0,0 +1,1611 @@
+    : QWizard(parent)
+{
+        //Set the name and size of the main window
+    this->resize(800,600);
```
Maybe scale to percentage of curent resolution if someone havs super high resolution

peterNordin commented on HopsanGUI/Dialogs/OptimizationScriptWizard.cpp:
```diff
@@ -0,0 +1,1611 @@
+    this->addPage(pParametersWidget);
+    this->addPage(pObjectiveWidget);
+    //this->addPage(pOutputWidget);
+    //this->addPage(pRunWidget);
```
leftover comments

peterNordin commented on HopsanGUI/Dialogs/OptimizationScriptWizard.cpp:
```diff
@@ -0,0 +1,1611 @@
+    QWizardPage *pSettingsWidget = new QWizardPage(this);
+    pSettingsWidget->setLayout(pSettingsLayout);
+    pSettingsWidget->setPalette(gpConfig->getPalette());
+    setAlgorithm(0);
```
What algorithm is 0, 
ok saw that its undefined, maybe should have that in the Ops enum.
'NoAlgorithm' or similar

robbr48 replied:
It does not mean undefined (there is an "++i" in setAlgorithm). 0 simple means use the first algorithm in the drop-down menu (which is currently Nelder-Mead).

peterNordin replied:
ok, I did not mean undefined as undefined variable, I saw that i++, I just mean that the magic value 0 is unclear.
I would be nice if there would be a enum value or named value that represent, "nothing chosen" or similar. 
But not such a big deal

peterNordin commented on HopsanGUI/Dialogs/OptimizationScriptWizard.cpp:
```diff
@@ -0,0 +1,1611 @@
+    mpParametersLayout->removeWidget(mpParameterRemoveButtons.at(i));
+
+    delete(mpParameterLabels.at(i));
+    delete(mpParameterMinLineEdits.at(i));
```
Here and in other places, maybe should use ->deleteLater(), in case events or signals are queued to these widgets

peterNordin commented on HopsanGUI/Dialogs/OptimizationScriptWizard.cpp:
```diff
@@ -0,0 +1,1611 @@
+
+    // Look in both local and global scripts directory in case they are different
+
+    QDir scriptsDir(gpDesktopHandler->getScriptsPath()+"HCOM/objFuncTemplates");
```
We should not assume that paths are returned with trailing /, add /HCOM

peterNordin commented on HopsanGUI/HcomHandler.cpp:
```diff
@@ -2297,9 +2297,9 @@ void HcomHandler::executeRunScriptCommand(const QString cmd)
-            gpOptimizationDialog->next();
-            gpOptimizationDialog->next();
-            gpOptimizationDialog->next();
+//            gpOptimizationDialog->next();
```
Leftover comments

peterNordin commented on HopsanGUI/OptimizationHandler.h:
```diff
@@ -95,6 +95,7 @@ class OptimizationHandler : public QObject
 
     friend class HcomHandler;           //! @todo Should probably not be friends
     friend class OptimizationDialog;    //! @todo Should probably not be friends
+    friend class OptimizationScriptWizard;    //DEBUG! REMOVE LATER /robbr
```
Leftover ?

peterNordin commented on HopsanGUI/Dialogs/OptimizationDialog.cpp:
```diff
@@ -61,1742 +62,237 @@
-    pSettingsLabel->setFont(boldFont);
+    // Main layout
+    QVBoxLayout *pLayout = new QVBoxLayout(this);
+    this->setLayout(pLayout);
```
Actually it does matter a little. In debug mode a warning is printed to stderr
QLayout: Attempting to add QLayout "" to OptimizationDialog "", which already has a layout

peterNordin commented:
The help button in the optimization dialog is not connected to anything

peterNordin commented:
The buttons in the optimization dialog is missing tool-tips (they have setText) but that is not the same

peterNordin commented:
On KUbuntu, the script wizard does not work for me, I do the typical Position servo example, choose complex rf, check all the plot boxes, choose P-gain set min max range. But then in the opjective function generator, I choose Step.out.value and PostionSensor.out.value -> minimize -> PROBLEM

The function selction combo box only show me two blank choices, that take up all of may vertical screen space, I can click one of them, to get either Overshoot or Lowest value, but clicking Add Function does nothing and I can get no further

robbr48 commented on HopsanGUI/Dialogs/OptimizationDialog.cpp:
```diff
@@ -61,1742 +62,237 @@
-        for(int i=0; i<optSettings.mParamters.size(); ++i)
+        QString output = "obj: ";
+        QString objStr;
+        //! @todo This is an error in the code, this must be solved, previous code assumed that i is in range of objects -> ASSERT failed in QVector
```
The check is actually redundant already, since it both has the error check and this line above:
if(i >= (int)objectives.size()) continue;
This error will only occur if someone calls the function with objectives and values vectors of different size, which is of course not allowed. And it will not cause a crash anyway, it would just ignore the excessive objectives.

robbr48 commented:
I cannot recreate the function selection error on Ubuntu 17.10. It searches for optimization scripts in two directories:
/home/username/Documents/Hopsan/Scripts/HCOM/objFuncTemplates
/home/username/Documents/Git/hopsan/bin//../Scripts/HCOM/objFuncTemplates
If both of these are empty or if the paths are interpreted wrong, the box will be empty. I'm not sure what will happen if there are erroneous HCOM files in the directory though.

However, I think this is a separate issue. I did not change this code in any way in this PR, so it most likely also existed before. I think we should create a new issue about it.

peterNordin commented:
I think the commit made tody is incorrect, it seemt to belong to some other branch and contain lots of stuff.

peterNordin commented:
I will test this on windows and see if that works better.
You are right that the problems I was having are old problems. (Will report issue)

peterNordin commented:
There is something odd with the last commit, it adds changes to the tools submodule, (fmu checkers).
But that submodule has no new commits in the main repo ?

peterNordin commented:
Seems like it has been reverted to an older commit, unintentional I assume

peterNordin commented:
Seems to work on Windows, much better usability now.

Found these issues, but likely separate issue
* When selecting overshoot having two output variables selected, addFunction silently does nothing. (should show some error, that I cant choose two variables)
* When enabling plots, they lack usefull information like parameter names and legend names. It mostly show the number 1 everywhere, but I think it did that before as well.

robbr48 commented:
I reverted the accidental subproject commit, and also deleted a leftover debug output.

Yes, the plots should contain more information and there should be better error handling when adding functions. I will create separate issues for this (#1737 and #1738), since these problems already existed before.
